### DISCUSSION TITLE
High performance image processing and display ?

### DISCUSSION THREAD

**Discussion by 4lch** [AvaloniaUI/Avalonia#12675] [Q&A]
Hi and thanks in advance for your time !

I am new to Avalonia and writing an image editor app that does fairly specialized image processing.
I am trying to figure out a way to have it be performant and keep the UI smooth, and unsure about how to go with it.

My current solution is to have the viewmodel for the image element hold a NetVips image for the processing, rendering it to a stream, then loading this stream into an Avalonia Bitmap for display. This certainly works, but can take a while, up to ~200-500ms per image for the original 24MP size, about 5-6ms for small thumbnails which should be used anyway. This conversion needs to happen every time a parameter is adjusted for the image processing.

Is there a generally accepted way for performant 2D image processing and display with Avalonia ? I have read about using Skia but am far from knowledgeable about it yet for example.

Thanks a lot for your time and help !

Best,
Adrien

**Comment by dbriard** [comment]
I don't know what a NetVips image is, but you can modify a WriteableBitmap directly, or you can use a Skia SKBitmap and render it using Avalonia Custom draw operation.

For processing, you can improve performance using multi-core, SIMD, and most important, process only what is visible onscreen.

**Reply by 4lch** [reply]
Thanks a lot for your reply ! NetVips is a C# port of Libvips, a high performance C image manipulation library that comes with pretty much all the tools I would need for this to work. That said, it looks like it is designed for a use-case where one loads an image, defines a chain of operations, and the image is loaded and modified only when this chain of operations is connected to an output (which would be a stream here). Afaik, there is no way to make it share memory with a common Bitmap format, since the operations are lazy and always the result of the entire pipeline of actions. The image data is only loaded and fed through the operations when connected to an output, so there is never an intermediate or original image to work with, but merely a reference to the source image data.

I have played around with the thumbnail size and it looks like it might not be too bad after all... With smaller thumbnails, the conversion time is somewhere around 5-8ms, including the time needed for the NetVips image to render to a stream, and the time needed to create the Avalonia image from that stream. This is not noticeable for the end user so would probably work fine here, letting me have a live preview that updates very quickly upon changing image correction parameters.

I will definitely have a look at Skia as you mention, as it seems to have some image manipulation abilities while being very quick to render apparently. This is something nice to pick up for graphics work anyway.

Thank you for the suggestion of using multi-core and SIMD as well ! I will try to implement it nicely so that nothing holds the main UI thread. In a typical use-case, all images would be displayed at the same time (about 40 at most) so there isn't much to gain from only processing the ones on screen, but I hope to make it so that the file is only ever converted and worked on when one of its parameters changes, so when there are no changes, the display Bitmap is kept in memory and used instead of recalculating it all the time.

I gotta say, this is very new to me, but the community has been great to learn Avalonia and related tips, so thanks a lot for your time and help !

**Comment by stevemonaco** [comment]
There's basically two approaches I know of.

1. Use `WriteableBitmap`. You can mutate existing pixel data within, so it's not as expensive as allocating a new image every time you update. I'd use this approach if NetVips' pixel data can't be directly rendered by SkiaSharp. I have a [demo here](https://github.com/stevemonaco/AvaloniaDemos/tree/master/RealTimeBitmapAdapter) that covers the approach.

2. Use `SKBitmap.InstallPixels` and write a custom View using `ICustomDrawOperation` to render it. If NetVips' pixel data is a common (eg. BGRA8888) format and the pixel data is contiguous, you can directly wrap the native array. Avalonia doesn't render `SKBitmap` directly, so you need to render that yourself. This eliminates the CPU memory copy and the allocation, so intuitively it _should_ be the faster approach. I have a [demo here](https://github.com/stevemonaco/AvaloniaDemos/tree/master/SkiaBitmapAdapter) of that specific approach.

I haven't benchmarked the two approaches and both use different data processing / render update integrations, so you can't really compare them. They're also both animated differently (hitting 60fps 1080p, but not much higher).

**Reply by 4lch** [reply]
Awesome, thank you very much for linking those examples !
These indeed look like nice solutions to this topic, to avoid duplicating memory and unnecessarily copying data. I will make sure to have a deeper look at the examples to see how to implement them.
I am considering moving to Magick.NET for image processing, since NetVips looked neat, but I have more experience with Magick, and it makes more sense to me for an image editing software to be able to retain the original images in memory, and not just references to them that require loading every time an output is requested... It seems like the solutions you mention should work just as well with Magick.NET, granted the conditions are met so definitely keeping that in mind.

Thanks again !
